Add description for photos [s6]

Finemas opened the pull request:
Add text view co cells in photo collection and try to prepare the model for adding this attribute.

Finemas committed:
Change for test environment

Finemas committed:
Change for test environment

Finemas committed:
Add note feature for photos [s6]

Finemas committed:
Add note feature for photos [s6]

Finemas committed:
Add hiding keyboard feature for VC [s6]

Finemas committed:
Prepare model for adding descriptions of photos [s6]

Finemas committed:
Prepare model for adding descriptions of photos [s6]

Finemas committed:
Prepare model for adding descriptions of photos [s6]

jakubskorepapoq commented on zScanner/ViewModels/NewDocument/NewDocumentPhotosViewModel.swift:
```diff
@@ -10,6 +10,21 @@ import UIKit
 import RxSwift
 import RxRelay
 
+class Picture: Equatable {
```
Change name to PageDomainModel, put into a separated file, on a correct place in the folder structure

jakubskorepapoq replied:
use struct, don't use class until necessary

Finemas replied:
It already exists so I added as a new attribute.

skorepak commented on zScanner/Coordinators/AppCoordinator.swift:
```diff
@@ -23,8 +23,8 @@ class AppCoordinator: Coordinator {
     
     // MARK: Navigation methods
     private func showSplashScreen() {
-        let viewController = SeaCatSplashViewController(coordinator: self)
```
We don't want to commit these changes. You can keep them on your local machine, but don't commit them

Finemas replied:
I did it so I wouldn't all the time when creating a new branch from dev make these changes again. So I'll put it in stashes.

skorepak commented on zScanner/Coordinators/NewDocumentCoordinator.swift:
```diff
@@ -136,19 +136,20 @@ class NewDocumentCoordinator: Coordinator {
     
     private static func steps(for mode: DocumentMode) -> [Step] {
-            return [.folder, .photos, .documentType]
+//            return [.folder, .photos, .documentType]
```
Don't commit testing changes

Finemas replied:
Sorry, my bad.

skorepak replied:
It's not resolved, I can still see it 👀

skorepak replied:
I like the TODO! It's just too vague description. I guarantee you that you will not now what "it will be done" means in couple of months. I've put so many todos like this myself and I was wondering later what the heck I meant by that. Please but more specific description on where the todo can be fixed/removed and how (what code to remove - in this case ucomment and delete the temporary code). 

- just be more specific, please. It's a good habit.

Finemas replied:
It makes sense, thank you for your advice.

skorepak commented on zScanner/DomainModel/PageDomainModel.swift:
```diff
@@ -13,6 +13,7 @@ struct PageDomainModel: Equatable {
     var index: Int
     var correlationId: String
     var relativePath: String
+//    var description: String
```
Generally, don't commit commented code. 
You can add the code here and put empty string somewhere, you can commit `// TODO:`. but don't commit commented code without any other comment when to uncomment it or why it is there.

skorepak commented on zScanner/DomainModel/PageDomainModel.swift:
```diff
@@ -40,8 +41,10 @@ struct PageDomainModel: Equatable {
-        self.image = image
+        
+        self.image = picture.image
+//        if let description = picture.description { self.description = description } else { self.description = "" }
```
don't make inline `if` it just looks bad. 
you can use `let description  = picture.description ?? ""` it's called "ternary operator".

Finemas replied:
I agree with you. Now it looks better.

skorepak commented on zScanner/Services/Network/API/NativeAPI.swift:
```diff
@@ -37,11 +37,12 @@ struct NativeAPI: API {
-            return
-        }
+    
+        let configuration = URLSessionConfiguration.default
```
don't commit testing changes

skorepak commented on zScanner/Services/Storage/Model/PageDatabaseModel.swift:
```diff
@@ -14,6 +14,7 @@ class PageDatabaseModel: Object {
     @objc dynamic var index = 0
     @objc dynamic var correlationId = ""
     @objc dynamic var relativePath = ""
+//    @objc dynamic var descript = ""
```
don't commit commented code. 
Don't make the name of the variable shorter. Any shortening can bring confusion. we have autocomplete and sometimes it's even working 😅

Finemas replied:
I know but it says that I cannot use "description" because it's already used. Should I use a different name e.g. notes but I must change it everywhere.
It says exactly this "Cannot override with a stored property 'description'", "Getter for 'description' with Objective-C selector 'description' conflicts with getter for 'description' from superclass 'Object' with the same Objective-C selector"

jakubskorepapoq replied:
Stupid Realm 🙄 Yeah, use notes, or userDesription or something. You don’t have to change it everywhere. That’s what the mapper is for. Just convert DatabaseModel.userDescription into DomainModel.description

Finemas replied:
Done but It gonna need to migrate and I have no idea how to do it.

skorepak commented on zScanner/SupportFiles/Extensions/String.swift:
```diff
@@ -41,4 +41,16 @@ extension String {
+    func isPIN() -> Bool {
+        if let number = Int(self) {
+            
+            if (number/10000000) <= 53 && length == 9 { return true }
```
Discussed elsewhere. Use more obvious validating logic.

Finemas replied:
It makes more sense now hopefully.

skorepak commented on zScanner/SupportFiles/Extensions/UIViewController.swift:
```diff
@@ -12,4 +12,14 @@ extension UIViewController {
         return view.safeAreaLayoutGuide
     }
+    
+    func hideKeyboardWhenTappedAround() {
```
The name is not clear. It's not describing what are you doing. 
Don't put condition into the function name. This function do: setup gesture recognizer for tap. What's gonna happen with keyboard.
So something like `setupKeyboardHandling` or `setupTapGesture` or something that.

skorepak commented on zScanner/ViewControllers/NewDocument/NewDocumentPhotosViewController.swift:
```diff
@@ -117,6 +117,7 @@ class NewDocumentPhotosViewController: BaseViewController {
     }
     
     private func setupView() {
+        self.hideKeyboardWhenTappedAround()
```
remove `self.`

skorepak commented on zScanner/Views/NewDocument/PhotoSelectorCollectionViewCell.swift:
```diff
@@ -84,3 +99,17 @@ class PhotoSelectorCollectionViewCell: UICollectionViewCell {
+            textView.textColor = UIColor.lightGray
+        }
+    }
+}
```
Use `textField.placeholder`

Finemas replied:
It's TextView, so should I change to TextField?

jakubskorepapoq replied:
Oh, I didn’t realise. Yeah, try to change it to textField and try out if it works fine. The simpler solution the better. :)

skorepak commented:
The branch s6 should start from dev, not from s4. I think you got a bit confused with multiple branches working on multiple tickets. You just need to get used to some culture :) 

Rebase your s6 branch on top of dev. If you don't know how to do that, reach out on slack.

Finemas committed:
Change Picture class to PageDomainModel struct [s6]

Finemas committed:
Change Picture class to PageDomainModel struct [s6]

Finemas committed:
Change Picture class to PageDomainModel struct [s6]

Finemas committed:
Update dismissing keyboard handler method

Finemas committed:
Update dismissing keyboard handler method

Finemas committed:
Change photo description to TextField [s6]

Finemas committed:
Fix mis click of keyboard handler

I put this method in the bad view controller. It should be in new document photos section because of photo description.

Finemas committed:
Fix mis click of keyboard handler

I put this method in the bad view controller. It should be in new document photos section because of photo description.

Finemas committed:
Add red bottom border on edit [s6]

Finemas committed:
Add inset for photo description

skorepak commented on zScanner/Coordinators/NewDocumentCoordinator.swift:
```diff
@@ -136,19 +136,19 @@ class NewDocumentCoordinator: Coordinator {
             .enumerated()
-            .forEach({ (index, image) in
-                let page = PageDomainModel(image: image, index: index, correlationId: newDocument.id)
+            .forEach({ (index,page) in
```
Check the formatting, missing space after the comma. It's a very small detail, but it will just look better :)

skorepak commented on zScanner/DomainModel/PageDomainModel.swift:
```diff
@@ -40,8 +41,12 @@ struct PageDomainModel: Equatable {
 extension PageDomainModel {
-    init(image: UIImage, index: Int, correlationId: String) {
-        self.init(id: UUID().uuidString, index: index, correlationId: correlationId, relativePath: "")
+    init(image: UIImage, index: Int, correlationId: String, description: String? = nil) {
```
The thing with structs is that they create default initializer for oyu with all the variables you have specified in the struct itself. If you write your own initializer inside the struct, you will override the default one. 

if you will write initializer in the extension, you will keep the default init and new one to it. That's what I'm using here. I'm creating a new init which will provide unique ID. 

You just extended it by providing empty string `""` into the initializer and then setting up `self.description` later.  (I didn't notice this before)

Another thing you are doing is that you are storing `String` but your initializer accepts `String?` - optional. Don't do that. 
Either store optional or accept non-optional string on the init. Like this, you can store nil but when you read it, you receive empty string, which is something we don't want. Initialize and store the same value. and use directlu the `self.init()` on lie 45 instead of injecting it later on lines 49 and 50. 

The `image` here is used because it's a computed property with hidden extra logic, so it is not present in the initializer.

skorepak commented on zScanner/SupportFiles/Extensions/UITextField.swift:
```diff
@@ -19,4 +19,41 @@ extension UITextField {
+            
+            self.rightView = paddingView
+            self.rightViewMode = mode
+        }
```
What is the purpose of all of this code?

Finemas replied:
I notice that the client sent screenshots how it should look like so I try to do it the same. I was a little bored, so I played with it a little. It just adds a red bottom border and indentation for photos descriptions.

skorepak commented on zScanner/SupportFiles/Extensions/UIViewController.swift:
```diff
@@ -12,4 +12,15 @@ extension UIViewController {
+    
+    @objc func dismissKeyboard() {
+        view.endEditing(true)
+    }
```
This code is used in one view controller only, why did you put it into UIViewController extension?

Finemas replied:
I don't even know. I'm sorry I'll put it to the specific UIVIewController.

Finemas committed:
Move keyboard handler to VC where is used

Finemas committed:
Update page domain model initializer

jakubskorepapoq commented on zScanner/Coordinators/AppCoordinator.swift:
```diff
@@ -9,53 +9,53 @@
-        let viewController = SeaCatSplashViewController(coordinator: self)
-        changeWindowControllerTo(viewController)
+ //        showSplashScreen()
+        startDocumentsCoordinator(with: UserSession(login: LoginDomainModel(username: "Test")))
```
Remove these duplicated changes

Finemas committed:
Clean a messy code

Finemas committed:
Merge two functions of bottom border [s6]

Finemas committed:
Add a border animation [s6]

skorepak committed:
Cleanup UITextField extension. Reorder parameters for cleaner usage. Simplify implementation. Set padding using constraints.

skorepak committed:
Fix collection view constraints

skorepak committed:
Add over-scroll to collection view. Enable to add description to images on the last row.

skorepak committed:
Unify behaviour of text fields on login screen with text fields on upload photos screen

skorepak committed:
Separate Page and PageDomainModel

Finemas committed:
Connect view and model of photo description

skorepak commented on zScanner/Views/NewDocument/PhotoSelectorCollectionViewCell.swift:
```diff
@@ -36,27 +39,43 @@ class PhotoSelectorCollectionViewCell: UICollectionViewCell {
         super.prepareForReuse()
         
-        image = nil
+        page = nil
```
you need to recycle dispose bag too. Take a look at `FormFieldTableViewCell`

skorepak commented on zScanner/Views/NewDocument/PhotoSelectorCollectionViewCell.swift:
```diff
@@ -75,6 +94,15 @@ class PhotoSelectorCollectionViewCell: UICollectionViewCell {
     
+    private lazy var textField : UITextField = {
+        let text = UITextField()
+        text.placeholder = "newDocumentPhotos.description.placeholder".localized
```
setup font explicitly to respect "Dynamic Type". Use one of the fonts specified in `UIFont.swift` extension

skorepak commented on zScanner/Views/NewDocument/PhotoSelectorCollectionViewCell.swift:
```diff
@@ -75,6 +94,15 @@ class PhotoSelectorCollectionViewCell: UICollectionViewCell {
+    private lazy var textField : UITextField = {
+        let text = UITextField()
+        text.placeholder = "newDocumentPhotos.description.placeholder".localized
+        text.adjustsFontSizeToFitWidth = true
```
I'm not sure this is a valid solution here. Make the placeholder text shorter instead.

skorepak commented on zScanner/Views/NewDocument/PhotoSelectorCollectionViewCell.swift:
```diff
@@ -86,14 +94,15 @@ class PhotoSelectorCollectionViewCell: UICollectionViewCell {
     }()
     
-    private var textField : UITextField = {
+    private lazy var textField : UITextField = {
```
You don't need to use lazy insede of UIView subclass, because the view is already created lazily. This is needed only on UIViewController because the view is needed later than the _instance_ of the View Controller itself. 

There is no obvious harm in doing that. If you take a look (or put a breakpoint) you will see it's called directlu in the init (trough `setupView()` ) and I believe it may slow down the compiler a little bit. if you use it only ince there is no issue, but if you'll have thousands of usages it will make the compile time longer. 

Again, this is just pointing out my experience and giving you more context. I'm perfectly fine with keeping the lazy keyword.

Finemas replied:
I also thought I didn't need to put it like lazy, but Xcode said that I can't use subscribe for the textfield without it. I was a bit confused.

Finemas replied:
Good point, thank you.

Finemas committed:
Remove temporary code

Finemas committed:
Prepare for reuse photo collection cell

Finemas committed:
Update text and font of photo description

Finemas committed:
Remove unnecessary code of photo description

Finemas committed:
Make code cleaner
